#!/usr/bin/env Rscript
#Rscript R02visual_minimap2_paf.R -i *.paf -o paf -A x.fai -B y.fai -a x.name -b y.name -m 5000 -n 50
#assemblyfasta=/mnt/content_176/yichun/pangenome/cicer_arietinum_hml/assembly/cicer_arietinum_hml.bp.p_ctg.fa
#reffasta=/mnt/content_176/yichun/pangenome/ref_genome/Cicer_arietinum_genomic.fa

#conda activate hifiasm
#echo "mapping ${reffasta} to ${assemblyfasta}."
#minimap2 -x asm5 -t 75 ${assemblyfasta} ${reffasta} > minimap.rassqref.paf
#Rscript /mnt/content_176/yichun/pangenome/Rscript/R02visual_minimap2_paf.R -i minimap.rassqref.paf -o paf -A cicer_arietinum_hml.bp.p_ctg.fa.fai -B Cicer_arietinum_genomic.fa.fai -a Cicer_arietinum_hml -b Cicer_arietinum_2013 -m 5000 -n 50

#echo "mapping ${assemblyfasta} to ${reffasta}."
#minimap2 -x asm5 -t 75 ${reffasta} ${assemblyfasta} > minimap.rrefqass.paf
#Rscript /mnt/content_176/yichun/pangenome/Rscript/R02visual_minimap2_paf.R -i minimap.rrefqass.paf -o paf -A Cicer_arietinum_genomic.fa.fai -B cicer_arietinum_hml.bp.p_ctg.fa.fai -a Cicer_arietinum_2013 -b Cicer_arietinum_hml -m 5000 -n 50

suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(optparse))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(plotly))

##Create parameters
option_list <- list(
  make_option(c("-i","--input"), type="character", default=NULL,
              help="'*.paf' generated by minimap2 [default %default]",
              dest="input"),
  make_option(c("-o","--output"), type="character", default="paf",
              help="output filename [default %default]",
              dest="output"),
  make_option(c("-A", "--xspecies-fai"), type="character", default=NULL,
              help="fasta index of reference (x) genome [default %default]",
              dest="fai_ref"),
  make_option(c("-B", "--yspecies-fai"), type="character", default=NULL,
              help="fasta index of query (y) genome [default %default]",
              dest="fai_query"),
  make_option(c("-a", "--xspecies-name"), type="character", default="x species",
              help="Reference name show at xlab [default %default]",
              dest="name_ref"),
  make_option(c("-b", "--yspecies-name"), type="character", default="y species",
              help="query name show at ylab [default %default]",
              dest="name_qry"),
  make_option(c("-m", "--min-alignment-length"), type="numeric", default=5000,
              help="minimum alignment length [default %default]",
              dest="min_alen"),
  make_option(c("-n", "--min-mapping-quality"), type="numeric", default=50,
              help="minimum mapping quality [default %default]",
              dest="min_mapq"),
  make_option(c("-p","--plot-size"), type="numeric", default=15,
              help="plot size X by X inches [default %default]",
              dest="plot_size"),
  make_option(c("-v", "--verbose"), action="store_true", default=TRUE,
              help="Print out all parameter settings [default]")
) 

options(error=traceback)

parser <- OptionParser(usage = "%prog -i *.paf  
                       -o paf -A x.fai -B y.fai -a x.name -b y.name -m 5000 -n 50  [options]",
                       option_list=option_list)
opt = parse_args(parser)

##manual add files
{
  #setwd("/mnt/content_176/yichun/pangenome/cicer_arietinum_hml/assembly/assembly_qc")
  #opt$input<-"minimap.rassqref.paf"
  #opt$output<-"paf"
  #opt$fai_ref<-"cicer_arietinum_hml.bp.p_ctg.fa.fai"
  #opt$fai_query<-"Cicer_arietinum_genomic.fa.fai"
  #opt$name_ref<-"Cicer_arietinum_hml"
  #opt$name_qry<-"Cicer_arietinum_2013"
  
  #setwd("/mnt/content_176/yichun/pangenome/cicer_arietinum_hml/assembly/assembly_qc")
  #opt$input<-"minimap.rrefqass.paf"
  #opt$output<-"paf"
  #opt$fai_ref<-"Cicer_arietinum_genomic.fa.fai"
  #opt$fai_query<-"cicer_arietinum_hml.bp.p_ctg.fa.fai"
  #opt$name_ref<-"Cicer_arietinum_2013"
  #opt$name_qry<-"Cicer_arietinum_hml"
}

##Read paf table
paf.table<-read.delim(opt$input, header = F)
names(paf.table)<-c("y.chr","y.len","y.start","y.end",
                    "strand", 
                    "x.chr","x.len","x.start","x.end",
                    "nmatch","alen","mapq","tp","cm",   
                    "s1","s2","dv","rl")

paf.table[c("y.len","y.start","y.end",
            "x.len","x.start","x.end",
            "nmatch","alen","mapq")] <- sapply(paf.table[c("y.len","y.start","y.end",
                                                         "x.len","x.start","x.end",
                                                         "nmatch","alen","mapq")],
                                               as.numeric)

##Convert y strand alignment
paf.table$y.start2<-NA
paf.table$y.end2<-NA
for (i in 1:nrow(paf.table)) {
  if (paf.table$strand[i]== "+") {
    paf.table$y.start2[i]<-min(paf.table$y.start[i],paf.table$y.end[i])
    paf.table$y.end2[i]<-max(paf.table$y.start[i],paf.table$y.end[i])
  } else {
    paf.table$y.start2[i]<-max(paf.table$y.start[i],paf.table$y.end[i])
    paf.table$y.end2[i]<-min(paf.table$y.start[i],paf.table$y.end[i])
  }
}

##Filter short alignments and bad mappings
paf.table<-paf.table[paf.table$alen >= opt$min_alen & 
                       paf.table$mapq >= opt$min_mapq,]
x.chrID<-unique(paf.table$x.chr)
y.chrID<-unique(paf.table$y.chr)

##Import chr/scaffold length and renumber
fai_ref<-read.table(opt$fai_ref, stringsAsFactors = F)
fai_ref<-fai_ref[,c(1,2)]
names(fai_ref)[1]="x.chr"
names(fai_ref)[2]="x.length"
fai_ref<-fai_ref[which(fai_ref$x.chr %in% x.chrID),]
fai_ref<-fai_ref[order(fai_ref$x.length, decreasing = TRUE),]
fai_ref$x.rank<-1:nrow(fai_ref)
row.names(fai_ref)<-1:nrow(fai_ref)
fai_ref$x.chrstart=NA
fai_ref$x.chrstart[1]<-0
fai_ref$x.chrend=NA
fai_ref$x.chrend[1]=fai_ref$x.length[1]
for (i in 2:nrow(fai_ref)) {
  fai_ref$x.chrstart[i]<-fai_ref$x.length[i-1]+fai_ref$x.chrstart[i-1]
  fai_ref$x.chrend[i]<-fai_ref$x.length[i]+fai_ref$x.chrend[i-1]
}
max.x<-max(fai_ref$x.chrend)

fai_query<-read.table(opt$fai_query, stringsAsFactors = F)
fai_query<-fai_query[,c(1,2)]
names(fai_query)[1]="y.chr"
names(fai_query)[2]="y.length"
fai_query<-fai_query[which(fai_query$y.chr %in% y.chrID),]

####order y axis
chr.freq<-as.data.frame(xtabs(~x.chr+y.chr, data = paf.table))
chr.freq<-chr.freq[order(chr.freq$y.chr,chr.freq$Freq, decreasing = TRUE),]
chr.freq<-chr.freq[which(chr.freq$x.chr %in% fai_ref$x.chr),]
row.names(chr.freq)<-1:nrow(chr.freq)

fai_query$x.chr<-NA
for (i in 1:nrow(fai_query)) {
  a<-subset(chr.freq, chr.freq$y.chr==fai_query$y.chr[i])
  a<-a[order(a$Freq, decreasing = TRUE),]
  row.names(a)<-1:nrow(a)
  fai_query$x.chr[i]<-as.character(a[1,1])
}
fai_query<-merge(fai_query, fai_ref[,c("x.chr","x.rank")], by = "x.chr", all.x = TRUE)
fai_query$y.rank<-NA
for (i in 1:nrow(fai_query)) {
  if (is.na(fai_query$y.rank[i])==TRUE) {
    b<-subset(chr.freq, chr.freq$x.chr==fai_query$x.chr[i] & chr.freq$Freq > 0)
    b<-b[order(b$Freq, decreasing = TRUE),]
    b$y.rate<-1:nrow(b)
    fai_query<-merge(fai_query, b[,c("x.chr", "y.chr", "y.rate")], by = c("x.chr","y.chr"), all.x = TRUE)
    if (is.na(fai_query$y.rank[i])==TRUE) {
      fai_query$y.rank[i]<-fai_query$y.rate[i]
      fai_query<-fai_query[,c(1:5)]
    } else {}
    
  } else{}
  
}

fai_query<-fai_query[order(fai_query$x.rank, fai_query$y.rank, decreasing = FALSE),]
#fai_query<-fai_query[order(fai_query$y.length, decreasing = TRUE),]
row.names(fai_query)<-1:nrow(fai_query)
fai_query$y.chrstart=NA
fai_query$y.chrstart[1]<-0
fai_query$y.chrend=NA
fai_query$y.chrend[1]=fai_query$y.length[1]
for (i in 2:nrow(fai_query)) {
  fai_query$y.chrstart[i]<-fai_query$y.length[i-1]+fai_query$y.chrstart[i-1]
  fai_query$y.chrend[i]<-fai_query$y.length[i]+fai_query$y.chrstart[i]
}
max.y<-max(fai_query$y.chrend)

cat(paste0("x length sum: ", max.x,"\n"))
cat(paste0("y length sum: ", max.y,"\n"))

##Chr/scaffold relengthed

##Assign new position to alignments
paf.table.filter<-paf.table
paf.table.filter<-paf.table.filter[,c("x.chr","x.len","x.start","x.end",
                                      "strand", 
                                      "y.chr","y.len","y.start2","y.end2",
                                      "nmatch","alen","mapq","tp","cm",   
                                      "s1","s2","dv","rl")]
names(paf.table.filter)<-c("x.chr","x.len","x.start","x.end",
                           "strand", 
                           "y.chr","y.len","y.start","y.end",
                           "nmatch","alen","mapq","tp","cm",   
                           "s1","s2","dv","rl")
####Merge x-reference
paf.table.filter<-merge(paf.table.filter, fai_ref[,c("x.chr","x.chrstart")], by = "x.chr", all.x = TRUE)
paf.table.filter$x.start2<-paf.table.filter$x.start+paf.table.filter$x.chrstart
paf.table.filter$x.end2<-paf.table.filter$x.end+paf.table.filter$x.chrstart

####Merge y-query
paf.table.filter<-merge(paf.table.filter, fai_query[,c("y.chr","y.chrstart")], by = "y.chr", all.x = TRUE)
paf.table.filter$y.start2<-paf.table.filter$y.start+paf.table.filter$y.chrstart
paf.table.filter$y.end2<-paf.table.filter$y.end+paf.table.filter$y.chrstart
##New (x,y) generated

##Plotting
##remove short contig labels
for (i in 1:nrow(fai_ref)) {
  if (fai_ref$x.length[i] <= max.x/40) {
    fai_ref$x.chr1[i]=""
  } else {
    fai_ref$x.chr1[i]<-fai_ref$x.chr[i]
  }
}

for (i in 1:nrow(fai_query)) {
  if (fai_query$y.length[i] <= max.y/40) {
    fai_query$y.chr1[i]=""
  } else {
    fai_query$y.chr1[i]<-fai_query$y.chr[i]
  }
}

##x and y tick mark
if (ceiling(max(fai_ref$x.chrend))>=250*10^6){
  breakseqx<-seq(0,
                 ceiling(max(fai_ref$x.chrend)/10^6),
                 ceiling(max(fai_ref$x.chrend)/(10^6*8*50))*50)
} else if (ceiling(max(fai_ref$x.chrend))>=100*10^6) {
  breakseqx<-seq(0,
                 ceiling(max(fai_ref$x.chrend)/10^6),
                 25)
} else {
  breakseqx<-seq(0,
                 ceiling(max(fai_ref$x.chrend)/10^6),
                 10)
}

if (ceiling(max(fai_query$y.chrend))>=250*10^6) {
  breakseqy<-seq(0,
                 ceiling(max(fai_query$y.chrend)/10^6),
                 ceiling(max(fai_query$y.chrend)/(10^6*8*50))*50)
} else if (ceiling(max(fai_query$y.chrend))>=100*10^6) {
  breakseqy<-seq(0,
                 ceiling(max(fai_query$y.chrend)/10^6),
                 25)
} else {
  breakseqy<-seq(0,
                 ceiling(max(fai_query$y.chrend)/10^6),
                 10)
}

cat(paste0("n alignments after filtering: ", nrow(paf.table.filter),"\n"))
##plot
p<-ggplot()+
  geom_point(data = paf.table.filter,
             aes(x=x.start2/10^6,
                 y=y.start2/10^6,
                 color = factor(x.chr)), size = 2)+
  geom_point(data = paf.table.filter,
             aes(x=x.end2/10^6,
                 y=y.end2/10^6), size = 2)+
  geom_segment(data = paf.table.filter,
    aes(x = x.start2,
        xend = x.end2,
        y = y.start2,
        yend = y.end2,
        #color = percentIDmean,
        text = sprintf(
          'x ID: %s<br>x Start Pos: %s<br>x End Pos: %s<br>y ID: %s<br>y Start Pos: %s<br>y End Pos: %s<br>Length: %s kb',
          x.chr, x.start2, x.end2,
          y.chr, y.start2, y.end2,
          round(alen / 1000, 1))))+
  scale_x_continuous(breaks = fai_ref$x.chrend,
                     labels = fai_ref$x.chr1,
                     sec.axis = sec_axis(~., name = "Mb",
                                         breaks = breakseqx*10^6,
                                         labels = breakseqx))+
  scale_y_continuous(breaks = fai_query$y.chrend,
                     labels = fai_query$y.chr1,
                     sec.axis = sec_axis(~., name = "Mb",
                                         breaks = breakseqy*10^6,
                                         labels = breakseqy))+
  labs(title = paste0("n=", nrow(paf.table.filter), 
                      " alignments, minimum alignment length=", 
                      opt$min_alen,
                      ", minimum mapping quality=",
                      opt$min_mapq),
       x = gsub("_"," ",opt$name_ref),
       y = gsub("_"," ",opt$name_qry))+
  geom_hline(yintercept = c(0,fai_query$y.chrend[fai_query$y.chr1 != ""]),
             linetype = "dotted", color = "gray50")+
  geom_vline(xintercept = c(0,fai_ref$x.chrend[fai_ref$x.chr1 != ""]),
             linetype = "dotted", color = "gray50")+
  theme(axis.line = element_line(linetype = "solid", size = 0.5),
        axis.ticks = element_line(colour = "black", size = 0.5),
        axis.title.x.bottom = element_text(size = 14, face = "italic"),
        axis.title.y.left = element_text(size = 14, face = "italic"),
        axis.title.x.top = element_text(size = 14),
        axis.title.y.right = element_text(size = 14),
        axis.text = element_text(size = 14, colour = "black"),
        axis.text.x.bottom = element_text(size = 14, angle = 90),
        axis.text.y = element_text(size = 14),
        plot.title = element_text(size = 14, vjust = 0.5),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 0),
        panel.background = element_rect(fill = NA),
        legend.key = element_rect(fill = NA),
        legend.background = element_rect(fill = NA),
        legend.position = "none")

p

filename=paste("minimap", opt$output, opt$name_ref, opt$name_qry, sep = ".")
ggsave(paste0(filename,".png"), width = opt$plot_size, height = opt$plot_size, units = "in", dpi = 300)
ggsave(paste0(filename,".tiff"), width = opt$plot_size, height = opt$plot_size, units = "in", dpi = 300)
gply = ggplotly(p, tooltip = "text")
htmlwidgets::saveWidget(as_widget(gply), file = paste0(filename, ".html"))

